test/test_utils/ec2.py

```diff
@@ -452,6 +458,44 @@ def get_ec2_instance_tags(instance_id, region=DEFAULT_REGION, ec2_client=None):
     return {tag["Key"]: tag["Value"] for tag in response.get("Tags")}
 
 
+def execute_asynchronus_testing_using_s3_bucket(
+    connection,
+    execution_command,
+    connection_timeout,
+    s3_location,
+    required_log_ending,
+    loop_time=2.5 * 3600,
+    loop_sleep_time=5 * 60,
+    log_location_within_ec2="~/container_tests/logs.txt",
+):
+    """
+    This method uses fabric to run the provided execution_command in asynchronus mode. While the execution command
+    is being executed in the image, it keeps on uploading the logs to the s3 bucket at fixed intervals. After a
+    loop_time is over, it checks the last line of the uploaded logs to see if it is same as required_log_ending.
+    This is mainly used in cases where Fabric behaves in an undesired way due to long living connections. 
+    :param connection: Fabric connection object
+    :param execution_command: str, command that connection.run() will execute
+    :param connection_timeout: timeout for fabric connection
+    :param s3_location: str, s3 location where the logs are supposed to be uploaded.
+    :param required_log_ending: str, The string that is desired to be present at the end of the logs
+    :param loop_time: int, seconds for which we would wait for the tests to execute on ec2 instance
+    :param loop_sleep_time: int, interval at which the logs would be uploaded to the s3 bucket
+    :param log_location_within_ec2: Location within ec2 instance where the logs are being witten.
+    """
+    connection.run(execution_command, hide=True, timeout=connection_timeout, asynchronous=True)
+    start_time = int(time.time())
+    loop_count = 0
+    while int(time.time()) - start_time <= loop_time:
```

Comment:
I'd suggest to add here something like ` or not last_line_of_log.endswith(required_log_ending)` into while condition. That will let you to stop execution immediately without waiting for the whole loop_time.

Plus maybe we need some sort of a check for hanged container. Something like "if no new lines appear during 15-30 minutes" - we drop the job and fail the test.

Reply:
This is a good idea. I will do it. Thanks!!